server/controllers/finance/reports/cashflow/index.js

```diff
@@ -753,3 +169,241 @@ function reportByService(req, res, next) {
     .catch(next)
     .done();
 }
+
+/**
+ * This function get periodic balances by transaction type
+ * reporting transaction type balance detailled by accounts
+ * with their balance for each transaction type
+ */
+function report(req, res, next) {
+  let serviceReport;
+  const dateFrom = new Date(req.query.dateFrom);
+  const dateTo = new Date(req.query.dateTo);
+  const options = _.clone(req.query);
+  const data = {};
+
+  // convert cashboxesIds parameters in array format ['', '', ...]
+  // this parameter can be sent as a string or an array we force the conversion into an array
+  const cashboxesIds = _.values(req.query.cashboxesIds);
+
+  _.extend(options, { orientation : 'landscape' });
+
+  // catch missing required parameters
+  if (!dateFrom || !dateTo || !cashboxesIds.length) {
+    throw new BadRequest('ERRORS.BAD_REQUEST');
+  }
+
+  try {
+    serviceReport = new ReportManager(TEMPLATE, req.session, options);
+  } catch (e) {
+    next(e);
+    return;
+  }
+
+  data.dateFrom = dateFrom;
+  data.dateTo = dateTo;
+
+  getCashboxesDetails(cashboxesIds)
+    .then(rows => {
+      data.cashboxes = rows;
+
+      data.cashAccountIds = data.cashboxes.map(cashbox => cashbox.account_id);
+
+      data.cashLabels = _.chain(data.cashboxes)
+        .map(cashbox => `${cashbox.label}`).uniq().join(' | ')
+        .value();
+
+      data.cashLabelSymbol = _.chain(data.cashboxes)
+        .map(cashbox => cashbox.symbol).uniq().join(' + ');
+
+      data.cashLabelDetails = data.cashboxes.map(cashbox => `${cashbox.account_number} - ${cashbox.account_label}`);
+
+      // get oposite lines to cashbox accounts in transactions as details
+      return getDetailsIdentifiers(data.cashAccountIds, dateFrom, dateTo);
+    })
+    .then(rows => {
+      data.detailsIdentifiers = rows.map(row => row.uuid);
+
+      // build periods columns from calculated period
+      return Fiscal.getPeriodsFromDateRange(data.dateFrom, data.dateTo);
+    })
+    .then(periods => {
+      data.periodDates = periods.map(p => p.start_date);
+
+      data.periods = periods.map(p => p.id);
+
+      data.colspan = data.periods.length + 1;
+
+      // skip period matrix if not transactions were identified
+      if (data.detailsIdentifiers.length === 0) { return []; }
+
+      // build periods string for query
+      const periodString = data.periods.length ? data.periods.map(periodId => {
+        return `SUM(CASE WHEN source.period_id = ${periodId} THEN source.balance ELSE 0 END) AS "${periodId}"`;
```

Comment:
This is correct and good SQL.  Just FYI - I like to use `IF()` because it makes the code more readable like this:
```js
const periodString = data.periods.length ? data.periods.map(periodId => {
  return `SUM(IF(source.period_id = ${periodId}, source.balance, 0)) AS ${periodId}`;
});
```

I know this comes from the database, but I would encourage you to always use `?` syntax to pass in parameters to SQL.  At some point, we'll make a mistake and have a terrible SQL injection error that [ends us up on international news](http://www.irinnews.org/news/2018/01/18/exclusive-audit-exposes-un-food-agency-s-poor-data-handling).